File: r/bob/bob.R
bob <- function(input) {
  input <- unlist(strsplit(trimws(input), ""))
  if (length(input) == 0) {
    "Fine. Be that way!"
  } else {
    is.a.question <- tail(input, n = 1) == "?"
    is.yelling <- length(input[grepl("[[:lower:]]", input)]) == 0 & length(input[grepl("[[:upper:]]", input)]) != 0
    if (is.a.question & is.yelling) {
      "Calm down, I know what I'm doing!"
    } else if (is.a.question) {
      "Sure."
    } else if (is.yelling) {
      "Whoa, chill out!"
    } else {
      "Whatever."
    }
  }
}

File: fsharp/bob/Bob.fs
module Bob

open System

let response (input: string) =
    let responseToSpeech (input: string) =
        let yelling input =
            let rec yellingYet input yet =
                match input with
                | [] -> yet
                | h :: t when Char.IsLower(h) -> false
                | h :: t when Char.IsUpper(h) -> yellingYet t true
                | _ :: t -> yellingYet t yet

            yellingYet (input |> Seq.toList) false

        let quizzical (input: string) =
            input.EndsWith('?')

        match (yelling input, quizzical input) with
        | (true, true) -> "Calm down, I know what I'm doing!"
        | (true, false) -> "Whoa, chill out!"
        | (false, true) -> "Sure."
        | (false, false) -> "Whatever."

    match input.Trim() with
    | "" -> "Fine. Be that way!"
    | speech -> responseToSpeech speech
